xarray/core/indexing.py

```diff
@@ -686,6 +686,10 @@ def _wrap_numpy_scalars(array):
     """Wrap NumPy scalars in 0d arrays."""
     if np.isscalar(array):
```

Comment:
In theory yes, but with that logic I get a failing test coming from `test_indexing.TestMemoryCachedArray.test_index_scalar`. `indexing.MemoryCachedArray` is not recognized as an array by `is_duck_array`.

Reply:
then we just need to figure out how to deal with that. Maybe check for classes that inherit from `xarray.core.utils.NDArrayMixin`?

Reply:
Could `MemoryCachedArray` be given an `__array_namespace__` attribute ?

Reply:
> Could MemoryCachedArray be given an __array_namespace__ attribute ?

Not yet.

Can we reorder these to check for dtype first and then isscalar?

Reply:
I don't think so.

Reply:
Assuming our test coverage is good enough, this appears to satisfy both your new test as well as all the existing ones (not sure if `NDarrayMixin` or `ExplicitlyIndexedNDArrayMixin` would be better to check):
```suggestion
if np.ndim(array) == 0 and (isinstance(array, np.generic) or not (is_duck_array(array) or isinstance(array, NDArrayMixin))):
```
From the table in the docstring of `numpy.isscalar`, `np.ndim(x) == 0` is `np.isscalar` plus 0d arrays and other python builtins / third-party objects. With the second part of the condition we exclude 0d arrays and our lazy indexing adapters.

Additionally, `is_duck_array` currently returns `True` for `numpy` scalars (new in `numpy=2.1`, I believe), so we have to explicitly include those again – this can be dropped after improving `is_duck_array`.